### ISSUE TITLE
Does Ovule.Nomad work with Mono on Linux?

### ISSUE THREAD

**Issue by wind39** [tony-dinucci/Ovule.Nomad#2]
Does it?

**Comment by BlackCentipede** [comment]
At the moment, it apparently doesn't. I will attempt to do some fixes and do a pull request after correcting some of the exceptions I've discovered here. It seems to throw exceptions like this line: NomadClient.cs: 293

![screen1](https://cloud.githubusercontent.com/assets/9369939/7214684/aaa82d0e-e573-11e4-9982-5a0390817e3a.png)

The "file://" and "file:" prefix seems to throw off Path.Combine method and create an incorrect path to files. Another part is the issue with the stock server that it initially run ok until it receive request from the client program and throw an exception originating from:

System.ServiceModel.Dispatcher.BaseRequestProcessor.ProcessRequest (System.ServiceModel.Dispatcher.MessageProcessingContext mrc)

with this message:

Exception Text content buffer exceeds the quota limitation at 636. 535896 bytes and should be less than 8192 bytes   at System.Xml.XmlBinaryDictionaryReader.Alloc (Int32 size) [0x00000] in <filename unknown>:0

I wouldn't know until I investigate more in depth with this, so I'll keep you posted.

**Comment by tony-dinucci** [comment]
Its not currently compatible with Mono I'm afraid. The main issue is that a manner of configuring WCF was used that's not in Mono. I didn't know this at time. Won't be hard to fix that one. Just got myself a new Linux box last week and intending on getting running on this before long.

Will keep you posted.

Also, thanks for looking into it Blackcentipede.

**Comment by BlackCentipede** [comment]
Seems you're right about that as this source suggested: http://www.mono-project.com/docs/web/wcf/

I am looking into some workaround for this however.

**Comment by tony-dinucci** [comment]
I think moving back to ServiceHost would be the way to go (what I'm doing superceeds that).  It'll mean things are slightly more clunky when service is hosted in IIS on Windows but won't be too bad.

This method here needs removed in particular Ovule.Nomad.Server.NomadWcfServer.Configure(ServiceConfiguration config)

It's a bit of a hacky way MS added to make configuration more consistent between self-hosted and web server hosted services.

Are you happy enough picking things up on the "Mono compatibility" front?  If so that would help a LOT.  I made a start on basic fault tolerance last night (through either retries or failover if there's a cluster) and if someone else tool the Mono stuff it'd mean I can get this done quicker ;)

**Comment by BlackCentipede** [comment]
I could try and improve the compatibility between Linux, Windows, OSX, and various platforms. :+1: I think I may be close to building a working code for Ovule.Nomad on Mono.

**Comment by tony-dinucci** [comment]
Fantastic :smile:

**Comment by wind39** [comment]
Thank you guys, you are awesome! :smile: I'm going to see if I can help with something.

**Comment by BlackCentipede** [comment]
Ok I think I got something here, part of the issue is the fact that Mono always have their security mode off for their WCF TCP Binding while Microsoft always had their on. When I set both security mode off, it suddenly work flawlessly between Mono (Client) and Windows (Server). If one side have encryption while other do not, you'll likely get this exception:

Ovule.Nomad.NomadException: Exception executing nomad method, see inner exception for more details: 'Preamble Ack Record is expected, got FFFFFFFF' ---> System.ServiceModel.ProtocolException: Preamble Ack Record is expected, got FFFFFFFF

I am still trying to make Mono Server to work with WIndows Client for WCF, anyhow, for now you can view my changes here:

https://github.com/BlackCentipede/Ovule.Nomad/commit/8f4d76c3d26eb821cad23297d7b9e06f56f56e4b

And you can see some of my test environment and changes I've made which worked out to this:

binding = new NetTcpBinding (SecurityMode.None);

All in all, I'm getting closer and closer to make a working copy of Ovule.Nomad compatible to Mono and .Net Framework. :+1:

**Comment by tony-dinucci** [comment]
Nice work!  
I'll get a better idea of the problem you're taking about in your edit once I'm able to pull your checkin (should be couple of hours).  Would be very nice to have a standard release that doesn't require any configuration on developers part to get up and running (for simple use at least) however won't be the end of the world if this isn't possible cross platform.  We should at least be able to do something where it's possible to get up and running on a single platform with no config but cross platform takes a little setup - which can be documented in Wiki.

William, the more help the better :smile: Let me know if there's a particular area that you're most interested in.

I suppose should really get a list of user stories together that cover what's going to be needed going forward.

**Comment by BlackCentipede** [comment]
Yeah, just that we don't have access to configuration for WCF Server on Mono side of things considering ServiceConfiguration wasn't implemented in Mono (I checked Mono source code btw.) So in this part, we get to be a little creative on how to solve this, ha ha.

**Comment by tony-dinucci** [comment]
If you're happy doing this then great.  Do you know if this stuff's actually available in one of the MS GitHub repos?

The ServiceConfiguration stuff is really just there to unify the way configuration (when done with code) is acheived with both hosted and self-hosted services and we should be able to get away without it, esp. if we focus on just self hosted services for the short term (this is easier to develop against, write samples and Wiki articles for and things are still a bit away from being release worthy).  https://msdn.microsoft.com/en-us/library/hh205277%28v=vs.110%29.aspx 

If it's not something you're overly keen on I think we could live with having self-hosted services just defaulting to TCP bindigs using normal ServiceHost for the moment when running under Mono.  If one side is Windows then at that side use a configuration file, i.e. standard WCF settings to tweak whatever defaults don't quite match up across platforms.  I don't know if I'm miles off here or not though - I want to give you're stuff a shot on my Linux machine tonight as there's no doubt stuff I'm missing.

BTW, just added a page to the Wiki here.  I'm open to debate (as with everything) but we need to have consistent coding standards otherwise merging is going to be hard. https://github.com/tony-dinucci/Ovule.Nomad/wiki/Development

Thanks

**Comment by BlackCentipede** [comment]
I'll fix the coding standards and do another pull request properly (sorry about late night pull-request.) Also it is one of my first time dealing with WCF and doing other way like you suggested may be possible. I'll give it a shot. :smile:

**Comment by wind39** [comment]
I'm ok if Nomad can't work with a Linux client and a Windows server, or vice-versa. It's ok if both the client and the server could run on Linux.

I'm particularly interested in the MapReduce and Load Distribution capabilities. I already tried to implement shared memory between process in Mono, but the following potential tools don't work in Mono:

- Memory Mapped Files (throw NotImplementedException)
- Mutex (doesn't lock between processes, only between threads)
- Named Pipes are not supported on Linux too.

I know Nomad uses WCF under the hood, but I never used WCF before. But WCF on Mono seems to be "limited": http://www.mono-project.com/docs/about-mono/compatibility/
So, maybe we really won't be able to run both Nomad server and client on Mono, at least for now.

See, I'm a Linux guy, and I use Mono for more than 3 years now. I never ran across such a mess in Mono before, every possible tool to solve my problem is incomplete in Mono.

But I'm really glad you guys are interested in making Nomad work in Mono. Please tell me how I can help.
I have a library called Spartacus (https://github.com/wind39/spartacus), it has Excel and PDF reporting capabilities. My ideas was to use inter process communication to be able to render huge PDF reports (~ 10000 pages) in a fastest way. Today, Spartacus can render 1000 pages in 2 min, but 10000 pages take 2 hours.
If Nomad could run on Mono, I could use Spartacus to test Nomad, what do you guys think?

**Comment by tony-dinucci** [comment]
Sounds live a very good idea - using Spartacus as a test.

I've not heard from BlackCentipede today so unsure what's going on his end (it's Sunday).

I'll take a look this evening at getting it running through Mono but won't check my changes in.  I'll send you a zip of the code.  That's assuming I get it working, I think it should be OK.

I'll be in touch later tonight.

**Comment by tony-dinucci** [comment]
OK I've got it working now (I'm running a Lubuntu laptop) - it's definitely slower with Mono and what I've done is a bit hacky but hopefully you can at least get it up and running.  There are a lot of optimisations I want to make anyway so will check what's causing the extra slowness on Mono at the same time (will do that over this coming week).  

As mentioned I've not checked anything in.  Will hopefully catch up with BlackCentipede later as he's already spent more time on this than I have - as said I've just done a hack job for now.

A tarball (containing sources and binaries) is here: https://www.dropbox.com/s/320cmyo4nj8eu0s/src.tar.gz?dl=0

I've duplicated the StockServer binaries 4 times and they're in on the path "Ovule.Nomad.Server.Stock/bin". 

I'm very interested to hear how things go with your test.  It sounds like you're planning on doing everything on the same machine - is that right?  When you're running an export at the moment do you still have plenty of resources spare to be running processes?

Thanks

**Comment by BlackCentipede** [comment]
Sorry about that, I had a lot of work for me today and finally made some time for this project, This sound like an interesting proposal for fixing the WCF issue. I've considered writing an alternative TCP/UDP network library project (respecting TAP programming pattern) to have the necessary flexibility in security and some room to expand different services in a single connection.

I'll recreate my fork and do another check on the source code. (I am still figuring out Git.)

**Comment by BlackCentipede** [comment]
I just looked at your new code... I can't believe I didn't find that method that would accept a custom TCP binding...

**Comment by wind39** [comment]
Oh my God, I'm so excited!! @tony-dinucci I tested your code, and IT WORKS!! :smile:

First I followed Hello World example, and it worked. So I tested MapReduce (I was very interested in it), and it also worked flawlessly.

So I tested the Load Distribution example. I tought it wouldn't work, because I tested many ways of sharing memory across processes in Mono, and all of them failed. But.. Load Distribution using your Nomad compilation WORKED!!

I can't say how thankful I am.. I'm anxious to write a real world example using my Spartacus' PDF report rendering features.
This week I worked hard, and today I completed a optimization in Spartacus' code. I reduced in 75% the execution time and in 33% the memory usage. Still there are some real world reports with ~15,000 pages which use more than 2GB RAM, and they still cause OutOfMemoryException.

My idea was to write a mix of MapReduce and Load Distribution, to distribute the creation of the render matrix between a number of Nomad Servers. The result need to be a single PDF file.
Yes, I'm going to do everything on the same machine. It is a Linux server with 20GB RAM. Maybe on the future I can make a test of rendering distributing the load in different machines, but now I have no resources.

There are SO MANY others use cases like this to be optimized with Nomad. Please, I will help in any way I can.

Just for the record:
- Debian 8
- Linux kernel 3.16.0-4
- Mono 3.12.1

**Comment by BlackCentipede** [comment]
Excellent! :+1: That's a great news! He should do a branch for Mono in this project with that code.

**Comment by tony-dinucci** [comment]
Great news! I'm so happy to hear that William.

BlackCentipede, no problem at all.  Yes I was thinking about getting a way to communicate setup (rather than purely through issues) too.  I was thinking it'd be good to have a forum or something where people who aren't neccessarily developing can ask questions too - I'm not sure what the best ones are (would be very good if it's hosted for free).  uTox might also be good - I think we might all be in very different timezones though so live chat would be limited, I'm UK, William is Brazil (at a guess I'm thinking you're maybe in Australia or something BlackCentipede?).

Alternative version of the server that work with plain TCP sockets would be great (as would UDP, probably RUDP would be better if there's a decent .Net library around for this).  I actually started doing this (plain TCP sockets) a while back but put it on the back-burner so that I wasn't holding up the main functionality too much.  I think this would be a very interesting and worthwhile project if you want to take it on.  Just raise a new issue if you want to take this on.

I'm at work just now but will merge your changes later when I get home.

Thanks again guys, It's exciting that things are starting to move forward!  BTW, I'll close this issue once the merge is done.

**Comment by BlackCentipede** [comment]
I also have started with that networking library as well: https://github.com/BlackCentipede/LightNet

Still prototyping that idea though. Btw, I'm in USA.

**Comment by tony-dinucci** [comment]
I see you've added you're name too - will be much easier to type Tyler than BlackCentipede all the time ;)

I've merged your pull request pull request in now and given it a test on Windows and things are looking good.  I'm still at work so can't check with Mono on Linux till later but sure it'll be fine.  I'll close off this issue once it's been confirmed everything that's in the repo runs on Linux.

Tonight I'll also see about setting up a bulletin board of some type so we're not using issues so much for general chat.

**Comment by wind39** [comment]
I suggest we can use Slack for chat (https://slack.com/). It basically group conversation by subject (channels).
It's free for an unlimited number of users/channels, and the free version have 5GB storage for attachments.

**Comment by tony-dinucci** [comment]
:+1: signed up for an account, invites should be on their way to you
